src/main/java/com/ldtteam/structurize/storage/ClientStructurePackLoader.java

```diff
@@ -0,0 +1,274 @@
+        FINISHED_LOADING,
+        SYNCING,
+        FINISHED_SYNCING
+    }
+
+    /**
+     * Set after the client finished loading the schematics.
+     */
+    public static volatile ClientLoadingState loadingState = ClientLoadingState.LOADING;
+
+    /**
+     * Called on client mod construction.
+     */
+    public static void onClientStarting()
+    {
+        final List<Path> modPaths = new ArrayList<>();
+        final List<String> modList = new ArrayList<>();
+        for (IModInfo mod : ModList.get().getMods())
+        {
+            modPaths.add(mod.getOwningFile().getFile().findResource(BLUEPRINT_FOLDER, mod.getModId()));
+            modList.add(mod.getModId());
+        }
+
+        final Path gameFolder = Minecraft.getInstance().gameDirectory.toPath();
+
+        Util.backgroundExecutor().execute(() ->
+        {
+            // This loads from the jar
+            for (final Path modPath : modPaths)
+            {
+                try
+                {
+                    Files.list(modPath).forEach(element -> StructurePacks.discoverPackAtPath(element, true, modList));
+                }
+                catch (IOException e)
+                {
+                    Log.getLogger().warn("Failed loading packs from mod path: " + modPath.toString());
+                }
+            }
+
+            // Now we load from the main folder.
+            try
+            {
+                final Path outputPath = gameFolder.resolve(BLUEPRINT_FOLDER);
+                if (!Files.exists(outputPath))
+                {
+                    Files.createDirectory(outputPath);
+                }
+                Files.list(outputPath).forEach(element -> StructurePacks.discoverPackAtPath(element, false, modList));
+            }
+            catch (IOException e)
+            {
+                Log.getLogger().warn("Failed loading packs from main folder path: " + gameFolder.toString());
+            }
+
+            Log.getLogger().warn("Finished discovering Client Structure packs");
+
+            loadingState = ClientLoadingState.FINISHED_LOADING;
+        });
+    }
+
+    @SubscribeEvent
+    public static void onWorldTick(final TickEvent.ClientTickEvent event)
+    {
+        if (event.phase == TickEvent.Phase.END)
+        {
+            if (Minecraft.getInstance().level != null && Minecraft.getInstance().level.getGameTime() % 20 == 0 && loadingState == ClientLoadingState.FINISHED_LOADING)
+            {
+                loadingState = ClientLoadingState.SYNCING;
+                Network.getNetwork().sendToServer(new NotifyServerAboutStructurePacks(StructurePacks.packMetas));
+            }
+            else if (Minecraft.getInstance().level == null && (loadingState == ClientLoadingState.SYNCING || loadingState == ClientLoadingState.FINISHED_SYNCING))
+            {
+                Log.getLogger().warn("Client logged off. Resetting Pack Meta and Reloading State");
+                loadingState = ClientLoadingState.LOADING;
```

Comment:
shouldn't this be `ClientLoadingState.FINISHED_LOADING` else we're reverting to the state we had before the client started

Reply:
Yes, this resets the loading entirely and starts over. The next call basically sets this to finished loading then

Reply:
I'm not seeing where it starts over then, since both conditions here require different loading states

Reply:
onClientStarting();